### PR TITLE
feat: support codec v1 for blob DA

### PR DESCRIPTION
**Pull request by Thegaram**
### Purpose or design rationale of this PR

Implement the codecv1 spec: https://www.notion.so/scrollzkp/4844-Specs-52490bfb413a4beaa37b764273840029

### PR title

Your PR title must follow [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) (as we are doing squash merge for each PR), so it must start with one of the following [types](https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#type):

- [ ] build: Changes that affect the build system or external dependencies (example scopes: yarn, eslint, typescript)
- [ ] ci: Changes to our CI configuration files and scripts (example scopes: vercel, github, cypress)
- [ ] docs: Documentation-only changes
- [x] feat: A new feature
- [ ] fix: A bug fix
- [ ] perf: A code change that improves performance
- [ ] refactor: A code change that doesn't fix a bug, or add a feature, or improves performance
- [ ] style: Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc)
- [ ] test: Adding missing tests or correcting existing tests

### Deployment tag versioning

Has `tag` in `common/version.go` been updated or have you added `bump-version` label to this PR?

- [ ] No, this PR doesn't involve a new deployment, git tag, docker image tag
- [X] Yes

### Breaking change label

Does this PR have the `breaking-change` label?

- [x] No, this PR is not a breaking change
- [ ] Yes

### CODE REVIEW

File: common/types/encoding/codecv1/codecv1.go [hunk 1]
```diff
@@ -0,0 +1,497 @@
...
+		}
+		dataBytes = append(dataBytes, chunkHash.Bytes()...)
+	}
+
+	dataHash := crypto.Keccak256Hash(dataBytes)
+	return dataHash, nil
+}
+
+// constructSkippedBitmap constructs skipped L1 message bitmap of the batch.
+func constructSkippedBitmap(batchIndex uint64, chunks []*encoding.Chunk, totalL1MessagePoppedBefore uint64) ([]byte, uint64, error) {
```
Comment (zimpha) [inline common/types/encoding/codecv1/codecv1.go]: this should be a common function used by both codec v0 and v1 right?

File: common/types/encoding/codecv1/codecv1.go [hunk 2]
```diff
@@ -0,0 +1,497 @@
...
+		bytes := num.Bytes()
+		padding := 32 - len(bytes)
+		copy(bitmapBytes[32*ii+padding:], bytes)
+	}
+
+	return bitmapBytes, nextIndex, nil
+}
+
+// constructSkippedBitmap constructs the 4844 blob payload.
+func constructBlobPayload(chunks []*encoding.Chunk) (*kzg4844.Blob, *kzg4844.Point, error) {
```
Comment (zimpha) [inline common/types/encoding/codecv1/codecv1.go:280]: this should be used by all following version, should we put it in `da.go` or somewhere else?

File: common/types/encoding/codecv1/codecv1.go [hunk 3]
```diff
@@ -0,0 +1,497 @@
...
+	}
+
+	hash := crypto.Keccak256Hash(dataBytes)
+	return hash, nil
+}
+
+// NewDABatch creates a DABatch from the provided encoding.Batch.
+func NewDABatch(batch *encoding.Batch) (*DABatch, error) {
+	// this encoding can only support up to 15 chunks per batch
+	if len(batch.Chunks) > 15 {
```
Comment (kunxian-xia) [inline common/types/encoding/codecv1/codecv1.go]: Should we declare 15 as a constant?

File: common/types/encoding/codecv1/codecv1.go [hunk 4]
```diff
@@ -0,0 +1,497 @@
...
+func constructBlobPayload(chunks []*encoding.Chunk) (*kzg4844.Blob, *kzg4844.Point, error) {
+	// the raw (un-padded) blob payload
+	blobBytes := make([]byte, 2*31)
+
+	// the number of chunks that contain at least one L2 transaction
+	numNonEmptyChunks := 0
+
+	// challenge digest preimage
+	// 1 hash for metadata and 1 for each chunk
+	challengePreimage := make([]byte, 16*32)
```
Comment (kunxian-xia) [inline common/types/encoding/codecv1/codecv1.go]: Should we replace 16 with `MAX_NUM_CHUNKS+1`?

File: common/types/encoding/codecv1/codecv1.go [hunk 5]
```diff
@@ -0,0 +1,497 @@
...
+		// challenge: compute chunk data hash
+		hash := crypto.Keccak256Hash(blobBytes[currentChunkStartIndex:])
+		copy(challengePreimage[32+chunkID*32:], hash[:])
+	}
+
+	// blob metadata: num_chunks
+	binary.BigEndian.PutUint16(blobBytes[0:], uint16(numNonEmptyChunks))
+
+	// challenge: compute metadata hash
+	hash := crypto.Keccak256Hash(blobBytes[0:62])
```
Comment (kunxian-xia) [inline common/types/encoding/codecv1/codecv1.go]: Should we declare 62 as `MAX_NUM_CHUNKS * 4 + 2`?

File: common/types/encoding/codecv1/codecv1.go (hunk 2 above)
Comment (Thegaram) [reply common/types/encoding/codecv1/codecv1.go:280]: Maybe. Let's refactor when we add a new version.

File: common/types/encoding/codecv1/codecv1.go (hunk 1 above)
Comment (Thegaram) [reply common/types/encoding/codecv1/codecv1.go]: Fixed in [9133ad4](https://github.com/scroll-tech/scroll/pull/1193/commits/9133ad422d9aa11c4ad35c8603e132bfe08488e8).

File: common/types/encoding/codecv1/codecv1.go (hunk 3 above)
Comment (Thegaram) [reply common/types/encoding/codecv1/codecv1.go]: Done in [f158118](https://github.com/scroll-tech/scroll/pull/1193/commits/f1581185883bb182ce3424b2157935693abd9bda).

File: common/types/encoding/codecv1/codecv1.go (hunk 4 above)
Comment (Thegaram) [reply common/types/encoding/codecv1/codecv1.go]: Done in [f158118](https://github.com/scroll-tech/scroll/pull/1193/commits/f1581185883bb182ce3424b2157935693abd9bda).

File: common/types/encoding/codecv1/codecv1.go (hunk 5 above)
Comment (Thegaram) [reply common/types/encoding/codecv1/codecv1.go]: Done in [f158118](https://github.com/scroll-tech/scroll/pull/1193/commits/f1581185883bb182ce3424b2157935693abd9bda).

File: common/types/encoding/codecv1/codecv1.go [hunk 6]
```diff
@@ -0,0 +1,450 @@
...
+
+var (
+	// BLSModulus is the BLS modulus defined in EIP-4844.
+	BLSModulus *big.Int
+
+	// BlobDataProofArgs defines the argument types for `_blobDataProof` in `finalizeBatchWithProof4844`.
+	BlobDataProofArgs abi.Arguments
+
+	// MaxNumChunks is the maximum number of chunks that a batch can contain.
+	MaxNumChunks int = 15
```
Comment (colinlyguo) [inline common/types/encoding/codecv1/codecv1.go:31]: Would this value change in this codec version?

**Review by colinlyguo — DISMISSED** [review]

File: common/types/encoding/codecv1/codecv1.go (hunk 6 above)
Comment (Thegaram) [reply common/types/encoding/codecv1/codecv1.go:31]: I don't think so. If we change it, we'll also need to change the provers, which will require another codec version.

**Review by zimpha — DISMISSED** [review]

**Review by kunxian-xia — DISMISSED** [review]

File: common/types/encoding/codecv1/codecv1.go (hunk 6 above)
Comment (colinlyguo) [reply common/types/encoding/codecv1/codecv1.go:31]: Got it.

**Comment by colinlyguo** [comment]
@Thegaram What about merging this PR? will use the implementation in rollup-relayer.

**Pull request merged**